UPGRADE-4.0.md

```diff
@@ -0,0 +1,66 @@
+# Upgrade from v3.x to v4.x
+
+If you have any questions regarding this upgrade process, please consult the [`examples`](https://github.com/terraform-aws-modules/terraform-aws-vpc/tree/master/examples/) directory:
+
+If you find a bug, please open an issue with supporting configuration to reproduce.
+
+## List of backwards incompatible changes
+
+- The minimum required Terraform version is now 1.0
+- The minimum required AWS provider version is now 4.x (4.35.0 at time of writing)
+- `assign_ipv6_address_on_creation` has been removed; use the respective subnet type equivalent instead (i.e. - `public_subnet_assign_ipv6_address_on_creation`)
+- `enable_classiclink` has been removed; it is no longer supported by AWS https://github.com/hashicorp/terraform/issues/31730
+- `enable_classiclink_dns_support` has been removed; it is no longer supported by AWS https://github.com/hashicorp/terraform/issues/31730
+
+## Additional changes
+
+### Modified
+
+- `map_public_ip_on_launch` now defaults to `false`
+- `enable_dns_hostnames` now defaults to `true`
+- `enable_dns_support` now defaults to `true`
+- `manage_default_security_group` now defaults to `true`
```

Comment:
Yes, so by default these resources will automatically get created the second you create a VPC
- default security group
- default route table
- default NACL

And you cannot delete these - either through the console, CLI, or Terraform. It ensures the minimum requirements for a VPC to exists. However, when they are "un-managed" or not captured in Terraform, they do not get any tags or names and just appear as orphaned/standalone resources that do not follow any conventions. To find out who or what they belong to,  you have to back track through the VPC association. 

In addition to the conventions (tags, names), it is strongly recommended to remove the default rules which are overly permissive. Again, they need to be there for the bare minimum VPC to exist, but almost all security scanners/tools/teams will flag these rules and require them to be removed. 

So with that, most users are enabling these to be "managed" (or they should be), and then the very next step they are doing is removing the default rules and instead using custom rules. We cannot do the 2nd part because that would be quite disruptive and we don't know what rules would over the broad spectrum for most users. However, we can reduce the number of steps by making the assumption here that these should be set to managed by default. What are your thoughts on this - keep as managed = true or revert to not managed?

Reply:
> What are your thoughts on this - keep as managed = true or revert to not managed?

Since this is going to be a breaking change (aka "major release"), we can leave it as `true`.

~~The only reason I can think of for why users may not want to have this feature enabled is to have backward compatibility or an inability to upgrade the version of this module.~~